pkg/agent/controller/networkpolicy/allocator.go

```diff
@@ -63,24 +78,79 @@ func newIDAllocator(allocatedIDs ...uint32) *idAllocator {
 	return allocator
 }
 
-// allocate allocates an uint32 ID if there's available, otherwise error is returned.
-// It will try to reuse IDs that have been released first, then allocate a new ID by
-// incrementing the last allocated one.
-func (a *idAllocator) allocate() (uint32, error) {
+// allocateForRule allocates an uint32 ID for a given rule if it's available, otherwise
+// an error is returned. It will try to reuse the IDs that have been released first,
+// then allocate a new ID by incrementing the last allocated one.
+func (a *idAllocator) allocateForRule(rule *types.PolicyRule) error {
 	a.Lock()
 	defer a.Unlock()
 
 	if len(a.availableSlice) > 0 {
 		var id uint32
 		id, a.availableSlice = a.availableSlice[0], a.availableSlice[1:]
 		delete(a.availableSet, id)
-		return id, nil
+
+		// Add ID to the rule and the rule to asyncRuleCache.
+		rule.FlowID = id
+		a.asyncRuleCache.rules.Add(rule)
+
+		return nil
 	}
 	if a.lastAllocatedID == math.MaxUint32 {
-		return 0, fmt.Errorf("no ID available")
+		return fmt.Errorf("no ID available")
 	}
 	a.lastAllocatedID++
-	return a.lastAllocatedID, nil
+
+	// Add ID to the rule and the rule to asyncRuleCache.
+	rule.FlowID = a.lastAllocatedID
+	a.asyncRuleCache.rules.Add(rule)
+
+	return nil
+}
+
+// forgetRule adds the rule to the async delete queue with a given delay.
+func (a *idAllocator) forgetRule(ruleID uint32) {
+	a.asyncRuleCache.deleteQueue.AddAfter(ruleID, asyncDeleteInterval)
```

Comment:
I started with individual fields in the `idAllocator` struct. Felt naming with separate struct is easier (probably a trivial reason) and also if someone wants to extend the cache usage to get the rules by a different key. 
If you think it is tidier, I can move back to the individual fields.

Reply:
My point was the struct doesn't have any encapsulation, its caller idAllocator just calls its member's methods directly, which leads long code like this line and `a.asyncRuleCache.rules.GetByKey(strconv.Itoa(int(ruleID)))`. And because there is no encapsulation, even if another struct wants to reuse asyncRuleCache, it doesn't save code when using it because the struct still needs to access asyncRuleCache's members and call their methods directly. For now, I feel they could just be individual fields for simplicity.

Reply:
Ok. Changed to individual fields and added appropriate names. Thanks.